### ISSUE TITLE
Contribute to django-helpdesk 0.3x

### ISSUE THREAD

**Issue by alligatorbait** [django-helpdesk/django-helpdesk#878]
Discussion to help work the 0.3x release and progress this project towards compatibility with the current Django version.

There are tasks to complete the 0.3x release, the purpose of this ticket is to have a place to discuss current goals and help newcomers know where they can put effort and pull requests to move support the next version release.

Would one of the main contributors please post some primary items that need attention that I might be able to work on?

Thanks.

**Comment by gwasser** [comment]
Thank you very much @alligatorbait ! I haven't had the time to work on this like I had hoped.

The `master` branch is the 0.3.x work. As far as I know it mostly needs to be updated for Django 3.x. There's some kind of bug in it that I possibly introduced when merging a bugfix from 0.2.x, but I haven't gone through to figure out where the error is.

You can also check the tickets under the 0.3.0 milestone. It's possible some of them should be deferred to 0.4.x, we need someone to check them and triage them and figure out only the minimum needed to get the 0.3.x release out.

Thanks!

**Comment by gwasser** [comment]
I didn't look too deeply, but it looks like many of the "bug" tickets on the 0.3.x milestone are related to Django 3.x support so I think that would be the top priority for the next release. Thanks again for your help!

**Comment by Benbb96** [comment]
Hi all !

I believe the first bug to fix is the import of an old function which doesn't exist anymore in master : `apply_query` (in `staff.py` line 30). Each time I run on master, I'm having trouble with that error. I can try to understand the new `__Query__` class and fix it in a PR.

At the same time, I will take a look into the things that are missing in order to make **django-helpdesk** running with the third version of Django. But as I'm not using it my other projects, I'm not sure I will manage to do it..

Then I would like to work on replacing some fields (ticket `description` and `resolution`, followup `comment` and preset-reply `body`) into HTML Fields with the use of the **TinyMCE** library in order to have better formatting and also keep the HTML from email (instead of having an attachment with _email_html_body_). See #859 for reference.

Another thing I would like to add is a **Merging** option on the ticket list : select some tickets and then, on an intermediate page, choose which one will receive all the followups of the other ones, also select which information for each ticket's fields. If an email is sent to a merged ticket, the followup has to be attached to the ticket to which it was merged.

Once the TinyMCE widget is integrated, I could add a way to **copy-paste screenshots/images** directly in it, in order to easily display image in an HTML answer. Images would be uploaded in background through AJAX as a Django media, and then the URL would be used in the HTML (URL has to be absolute since it will be used in emails).

One last thing I would like to try to implement  is the adding of **reports on custom fields** (the list one in particular).
For exemple, if I add a custom fields with a billing information as a list (maintenance contract, quote, incorrect billing, offered), then I would like to know how many tickets are in which billing per user, or per queue.

I already have implemented some of these features on my fork that I'm using in my company project and I think they can be useful to others. This would be part of my first participation to **HacktoberFest** 😄

**Comment by alligatorbait** [comment]
The apply_query issue was introduced back in [Refactor query code into customizable class](https://github.com/django-helpdesk/django-helpdesk/commit/b96d725239f1e69eb8b0e968ea8095f712c79404) where one instance of apply_query() was missed in a refactor and has resided in the code since. The item that needs to be updated to Query() is in run_report() in staff.py.

Then in #850 [from timthelion/patch-1](https://github.com/django-helpdesk/django-helpdesk/commit/6463fbdcd9a58d276e2ca66b5206544e4dcc3fd3) the reference to the non-existent apply_query() was also added back into staff.py.

Anyone want to volunteer to update that first apply_query() over to the new Query() class object in run_report() in staff.py and remove the reference to import the apply_query() from staff.py at the same time.

**Comment by gwasser** [comment]
Thanks @Benbb96! This is a great list of items, I hope it's not too much for one month but will certainly be a big improvement if so! I think we'd have a solid 0.3.x release if we got that done and fixed the errors like @alligatorbait mentioned.

**Comment by alligatorbait** [comment]
@Benbb96 and @gwasser, I used the same apply_query change @Benbb96 made and then went and made all the other changes in order to have django-helpdesk run with Django 3.1.2 (current version as of this message).

**Comment by alligatorbait** [comment]
@Benbb96 you have me looking at the Hacktoberfest shirt now that I know about this.

**Comment by gwasser** [comment]
It occurs to me that the travis ci was still using django 2.x for testing, so I updated the travis ci config. I'll merge the two PRs so we don't lose anything and see where we're at.

**Comment by gwasser** [comment]
So there appears to still be some error still happening but I think we're close! I noticed in the build log we get errors like:
`'saved_queries' template tag (django-helpdesk) crashed with following error:

'AnonymousUser' object is not iterable

.'saved_queries' template tag (django-helpdesk) crashed with following error:

'AnonymousUser' object is not iterable`

**Comment by Benbb96** [comment]
@gwasser This error can be fixed in `templatetags/saved_queries.py` by adding a condition in the beginning of the function : 

    if user.is_authenticated:

But I didn't have time to investigate for the other errors in the travis build.

**Comment by alligatorbait** [comment]
I entered a merge request to correct the is_authenticated issue and then found additional test errors that appear when running the tests locally.

**get_emails.py was updated but the tests were not**, and several errors appear due to this change:
1. **imaplib module missing** - This used to exist in get_emails.py and was imported from there. It looks like the whole test_get_email.py needs rework.
test_get_email.py", line 265, in test_commas_in_mail_headers

2. **list_dir module missing**
test_get_email.py", line 227, in test_commas_in_mail_headers

3. **poplib module missing**
test_get_email.py", line 249, in test_commas_in_mail_headers

**Comment by gwasser** [comment]
Thanks @alligatorbait! I bet the mail libraries in django got renamed and/or moved for Django 3. We'll have to update the imports as necessary.

Which brings a question: should `django-helpdesk` 0.3.x support Django 2.2+, or should we require Django 3? If both Django 2.2 and Django 3, then we need to be careful with conditional imports, otherwise if just Django 3 then we can replace with correct imports.

**Comment by gwasser** [comment]
I'm a little hesitant to not support Django 2.2 since it is an LTS release, and Django 3.2 LTS is still a good 6 months away (April 2021). But we could decide to simply be cutting edge and support only Django 3. What do y'all think?

**Comment by alligatorbait** [comment]
Yes, we should retain Django 2.2 support as well as support 3.2. Are you able to update the travis-ci options to do Django 2.2 builds in conjunction with the Django Latest version builds?

I will see about fixing these test_get_email errors, and then also look at the templates to see if I can find a way to provide one that can easily be overridden for a top level site theme.

**Comment by gwasser** [comment]
Thanks @alligatorbait! I updated travis to test Django 2.2 and Django 3.x latest. I think it would be good to support both Django 2.2 LTS and eventually 3.2 LTS so there's a clear upgrade path for folks.

**Comment by alligatorbait** [comment]
During runtime I see this file as missing.
"GET /static/helpdesk/helpdesk-customize.css HTTP/1.1" 404 1719

1. Should there be a place holder for this file?
2. I don't see any reference to this item in the documentation. I'm adding an html template customization in my next pull request and documenting it, should I add documentation for this at the same time?

`docs/custom_templates.rst:6:In particular, users can include a file named `helpdesk-customize.css` in their django project directory to provide CSS overrides easily.`
`helpdesk/static/helpdesk/helpdesk-extend.css:4:@import 'helpdesk-customize.css';`

**Comment by gwasser** [comment]
The `helpdesk-customize.css` file is probably unnecessary today. It's a long ago workaround for better customization, but django already supports overriding templates and such.

**Comment by alligatorbait** [comment]
Should we remove the support for the helpdesk-customize.css, it is not documented (at least not displayed). I can put in another pull request to remove that if we want.

**Comment by gwasser** [comment]
I honestly don't know the best way to support custom CSS in django, if it's all handled by modern django or if we need to do something in our templates to support it. If anyone knows or can do a little research, I support whatever the consensus is.

**Comment by alligatorbait** [comment]
I went through everything tagged with `0.3.0` and there are several excellent improvement suggestions, some bugs, and some items that have been around forever. But in the list I see no deal breakers to prevent a release of version 0.3.0 at this point. I would want to have a couple of people actually test functionality on both Django 2.2 and Django 3.1 and run through as many functions as possible to **make sure there are no deal breaker runtime errors**.

**Comment by alligatorbait** [comment]
In requirements.txt when the requirements are installed this error pops up:
  Cloning https://github.com/pinax/pinax-teams.git (to revision dd75e1c) to /tmp/pip-download-jrjkdds3/pinax-teams
  Did not find branch or tag 'dd75e1c', assuming revision or ref.

We should probably fix that.

**Comment by timthelion** [comment]
I haven't tested the changes yet, but I've rrad through all the PRs and it looks like some great work is going on. I did review most of the code, but didn't comment just because I didn't see anything obviously wrong 🙂👍

**Comment by alligatorbait** [comment]
My first run with current version gives me a module missing `markdown_deux`
I'll look at this and see what is causing it later today.

**Comment by gwasser** [comment]
I'm not familiar with pinax actually, that's part of the work from @timthelion toward better permissions management. We should investigate. I'm not a fan of requiring some random commit. If pinax has a new release, let's use that, if it's not seeing much development then I'm tempted to cut out the pinax dependency altogether and look for a different solution.

I feel like I looked into `markdown_deux` once before and it was an old package that has a replacement. I think we just need to drop in the new dependency instead.

**Comment by alligatorbait** [comment]
Actually, looks like someone already removed Markdown-deux, I'm seeing it on the readthedocs master, but in the repo latest on master no reference to it.

**Comment by gwasser** [comment]
That would explain why I thought I looked into it before! I wonder if readthedocs is stuck on an old revision. We used to use `master` as the current release with a `devel` branch that was the next release. I changed it to the more standard release branch with `master` tracking bleeding-edge development. I'll have to log in to readthedocs and see if anything needs changing.

**Comment by gwasser** [comment]
OK I just uploaded the 0.3.0b1 to github and pypi. Please test and keep reporting bugs and submitting PRs. Thank you!

**Comment by gwasser** [comment]
0.3.0 beta 3 is now tagged on github, uploaded to pypi and available on the demo website: https://django-helpdesk-demo.herokuapp.com. This fixes a few issues notably an issue with the login and translations. Please test it and send feedback and PRs! hoping we can turn this into a real release very soon.

**Comment by MrPiethon** [comment]
Since master is using Bootstrap v4.5.3 can I slowly start working on bumping this up to v5 (currently in beta)? v5 offers some really convenient form controls like [Datalists](https://getbootstrap.com/docs/5.0/forms/form-control/#datalists) amongst other things that would really provide some QoL.

**Comment by gwasser** [comment]
Sure go ahead. We've got to iron out some bugs in the 0.3.x release anyway
so doing cleanups for bootstrap 5 might help clean up the templates.

**Comment by MrPiethon** [comment]
@gwasser, Could I also improve the designs a bit so they look more modern and consistent with other help desks?

An example of a KB item template edit I made:

![image](https://user-images.githubusercontent.com/25252033/107313416-536ff300-6a47-11eb-80c0-ce6aa67df87d.png)

If so, is there any place in particular you would like me to share the designs?

**Comment by gwasser** [comment]
yeah that looks great! just replace the existing templates with new ones on
a pull request branch. as long as we're not losing features, ok with an
update.

**Comment by MrPiethon** [comment]
Awesome happy to hear that. Will start working on the PR once I finish the designs and release them on my platform.

**Comment by alligatorbait** [comment]
@MrPiethon I had updated the Create Ticket Buttons on the Knowledge Base section to display something other than 'Contact a Human', hoping you'll keep with the updated theme for the ticket creation buttons.

**Comment by alligatorbait** [comment]
Any objections to adding a DateTime picker? https://github.com/xdan/datetimepicker MIT license and active with significant utilization.

**Comment by MrPiethon** [comment]
> 
> 
> @MrPiethon I had updated the Create Ticket Buttons on the Knowledge Base section to display something other than 'Contact a Human', hoping you'll keep with the updated theme for the ticket creation buttons.

Are your changes present in master branch?

**Comment by alligatorbait** [comment]
Yes, basically the button text said "Create Ticket in Queue: xxxx xxxx"

**Comment by alligatorbait** [comment]
![image](https://user-images.githubusercontent.com/60234929/107568865-da86ad80-6ba4-11eb-890d-f39ceb76e617.png)
![image](https://user-images.githubusercontent.com/60234929/107568902-e70b0600-6ba4-11eb-8f8c-36fb1dd00187.png)

**Comment by alligatorbait** [comment]
I noticed that the change also broke a test that was looking for 'Contact a human'

```
FAIL: test_kb_category (helpdesk.tests.test_kb.KBTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/django-helpdesk/django-helpdesk/helpdesk/tests/test_kb.py", line 50, in test_kb_category
    self.assertContains(response, 'Contact a human')
  File "/home/travis/virtualenv/python3.6.12/lib/python3.6/site-packages/django/test/testcases.py", line 454, in assertContains
    self.assertTrue(real_count != 0, msg_prefix + "Couldn't find %s in response" % text_repr)
AssertionError: False is not true : Couldn't find 'Contact a human' in response

```

**Comment by gwasser** [comment]
Alright folks, thank you for all your contributions and testing! I think the 0.3.0 beta has been out long enough for testing and not too many recent bug reports on 0.3.0. I am inclined to release it publicly. Let's set a target date for the end of the month. Any cleaning up or merges or fixes we should do before making it public?

**Comment by alligatorbait** [comment]
It looks like the docs are updated, but when accessing from readthedocs the updated versions are not displayed. We will want that to update properly before the release.

**Comment by brendanwood** [comment]
Any update on this? I'd like to help, but I'm not familiar enough with the project to know what might be missing.

**Comment by gwasser** [comment]
@brendanwood have you used the beta? Any last minute bugs or problems? It's really just waiting on me to officially release it I think, which I can do this week. I've had a lot on my plate recently so apologize about the delay.

**Comment by brendanwood** [comment]
@gwasser I have not used the beta, but we're at the point where we may switch our prod version to the beta anyway. In fact, if that would help you decide whether beta is ready for a release, we can just do that and report back after a couple weeks of using it in prod.

I'll do that and let you know how it goes.

**Comment by gwasser** [comment]
Thanks! Yes I've been hopeful to get more feedback before officially
releasing it. If you plan to switch and don't mind reporting bugs, that
would definitely help get it officially released. Let me know your
experience switching over!

**Comment by brendanwood** [comment]
We have upgraded our test environment so far, we will roll this out to prod with our next deploy as long as everything continues to look fine. There were a few hurdles to upgrade our test environment, may need to update documentation so that others can upgrade smoothly.

- Under `INSTALLED_APPS`, `bootstrapform` needs to be replaced with `bootstrap4form`

- `pinax_teams` appears to be a new required dependency, but is not listed in `requirements.txt`. Is this correct?

- Assuming `pinax_teams` is a legit new dependency, need to add the following to `INSTALLED_APPS` for `pinax_teams`:

```
        "account",
        "pinax.invitations",
        "pinax.teams",
        "reversion",
```

- If using `send_templated_mail`, then it now needs to be imported from `helpdesk.templated_email`

- Missing migration for `Alter field email_on_ticket_change on usersettings`. Here is the migration file I generated. I can submit that in a PR if needed.

```
        # Generated by Django 3.1.13 on 2021-08-18 14:30

        from django.db import migrations, models
        import helpdesk.models


        class Migration(migrations.Migration):

            dependencies = [ 
                ('helpdesk', '0034_create_email_template_for_merged'),
            ]   

            operations = [ 
                migrations.AlterField(
                    model_name='usersettings',
                    name='email_on_ticket_change',
                    field=models.BooleanField(default=helpdesk.models.email_on_ticket_change_default, help_text="If you're the ticket owner and the ticket is changed via the web by somebody else,do you want to receive an e-mail?", verbose_name='E-mail me on ticket change?'),
                ),
            ]   
```

Other than that, the migration went smoothly so far. I will update once we push to prod.

**Comment by gwasser** [comment]
thank you @brendanwood for the update! this sounds promising that there were no major hiccups so far. I can make the changes you listed above and if things go smoothly I'll release this as 0.3.0. thanks for your help!

**Comment by gwasser** [comment]
FYI @brendanwood I made the changes you suggested, added the migration and added a new upgrade.rst file to the docs that give info on how to upgrade from older versions of django-helpdesk. Let me know if you have any other changes, or PRs welcome! In general, PRs welcome from all the folks in this thread, if you have time to tackle a few 0.3.0 bug reports before we release.

**Comment by brendanwood** [comment]
@gwasser Thanks for your patience on this. It took us longer than expected to get this rolled out to prod because there were a cascade of dependency updates from our Django3 migration.

We have been running the master branch of django-helpdesk in prod for the last week. Overall it seems quite stable.

We did encounter one major issue regarding email processing from Outlook-based email addresses. Here is a PR that appears to fix the issue for us:

https://github.com/django-helpdesk/django-helpdesk/pull/974

The error we were seeing was thrown by the postmark integration but caused by bad whitespace characters in certain message fields that was not being stripped. It's not clear whether this will affect other email providers, but it's important to make sure this is fixed before cutting the 0.3.0 release.

**Comment by gwasser** [comment]
Thanks for testing and sending a PR @brendanwood! I accepted it into `master`, should get rolled up into a beta shortly. I know of maybe one other main fix we should do before official release, but almost there. Anyone else following this thread: any last minute fixes? Bugfix PRs welcome!

**Comment by nickmccullum** [comment]
@gwasser wondering what the timeline looks like for an official release? We have been running the master branch in production for about 2 weeks now.

**Comment by gwasser** [comment]
There's a couple bugfixes I want to get in before cutting an official release, but I'd expect probably next week @nickmccullum. I'll probably roll up one last beta this weekend, wait a few days, then cut the release next week. So, last call for folks in the thread to point out any blocker or important bug, otherwise we'll put it off to 0.3.x bugfixes or 0.4 development branch.

**Comment by gwasser** [comment]
I uploaded a beta 5 just now, please give it one last look over this week. If no major blockers, I'll make it official 0.3.0 release by next Monday probably. Thanks everyone!
